### PR TITLE
Support for the new Linter API

### PR DESCRIPTION
**Pull request by Arcanemagus**
Upgrade to the new Linter API. Fix a few things with the rest of the project such as adding a CHANGELOG.md and a few minor issues with the readme.

### CODE REVIEW

**Comment by Arcanemagus** [comment]
Looks like a different version of this work was already done in https://github.com/AtomLinter/linter-php/pull/47.

I would say mine is more complete, but then I'm biased, so I'll leave that up to the reviewer to decide. :stuck_out_tongue:

**Comment by steelbrain** [comment]
Looks quite well, except that you are not using `atom-helper` npm module. Use `Helpers::exec` instead of `child_process.exec`

**Comment by Arcanemagus** [comment]
I'm using its `parse` method, I was having issues with `exec`, I'll re-diagnose them and detail them shortly. Maybe you can figure out what I was doing wrong.

**Comment by Arcanemagus** [comment]
So if I change the linter block to this:
```coffee
lint: (textEditor) =>
        # return new Promise (resolve, reject) =>
        data = []
        filePath = textEditor.getPath()
        command = @executablePath
        parameters = @parameters.filter (item) -> item
        parameters.push('--syntax-check')
        parameters.push('--no-php-ini')
        parameters.push('--define display_errors=On')
        parameters.push('--define log_errors=Off')
        parameters.push("--file '" + filePath + "'")
        return helpers.exec(command, parameters).then (result) =>
          return helpers.parse(result, @regex, {filePath: filePath})
```

Inside helpers.coffee the arguments look correct:
```
spawnedProcess.spawnargs
["php", "--syntax-check", "--no-php-ini", "--define display_errors=On", "--define log_errors=Off", "--file 'C:\Users\Landon\Documents\GitHub\plexWatchWeb\settings.php'"]
```

But the output given back is simply the php usage output, so something isn't being sent properly. I'm not sure how to get the full command it is executing, beyond writing a wrapper script that logs it and calling that as "php".

Any ideas?

File: lib/main.coffee [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    provider =
+      grammarScopes: ['text.html.php', 'source.php']
+      scope: 'file'
+      lintOnFly: false
+      lint: (textEditor) =>
+        return new Promise (resolve, reject) =>
+          data = []
+          filePath = textEditor.getPath()
+          command = @executablePath
+          parameters = @parameters.filter (item) -> item
```
Comment (steelbrain) [inline lib/main.coffee]: `@parameters` is un-used and redundant, You can just remove it

**Comment by steelbrain** [comment]
You are passing the filePath incorrectly, it should be
```js
parameter.push('--file', filePath)
```

**Comment by Arcanemagus** [comment]
This doesn't change anything unfortunately:
```coffee
parameters = []
# ...
parameters.push('--file', filePath)
```

**Comment by steelbrain** [comment]
I'll clone your branch locally and try to make it work.

**Comment by steelbrain** [comment]
@Arcanemagus here you go buddy
```diff
diff --git a/.gitignore b/.gitignore
index e69de29..2d2b47d 100644
--- a/.gitignore
+++ b/.gitignore
@@ -0,0 +1,2 @@
+.idea
+node_modules
\ No newline at end of file
diff --git a/lib/main.coffee b/lib/main.coffee
index 5923956..3d9c224 100644
--- a/lib/main.coffee
+++ b/lib/main.coffee
@@ -8,7 +8,6 @@ module.exports =
       default: 'php' # Let OS's $PATH handle the rest
 
   activate: ->
-    @parameters = []
     @regex = '(Parse|Fatal) ' +
             '(?<error>error):(\\s*(?<type>parse|syntax) error,?)?\\s*' +
             '(?<message>(unexpected \'(?<near>[^\']+)\')?.*) ' +
@@ -31,19 +30,13 @@ module.exports =
       lintOnFly: false
       lint: (textEditor) =>
         return new Promise (resolve, reject) =>
-          data = []
           filePath = textEditor.getPath()
           command = @executablePath
-          parameters = @parameters.filter (item) -> item
-          parameters.push('--syntax-check')
-          parameters.push('--no-php-ini')
-          parameters.push('--define display_errors=On')
-          parameters.push('--define log_errors=Off')
-          parameters.push('--file "' + filePath + '"')
-          command += ' ' + parameters.join(' ')
-          process = child_process.exec(command)
-          process.stdout.on 'data', (d) -> data.push d.toString()
-          process.on 'close', =>
-            messages = helpers.parse(data.join(''), @regex,
-                                     {filePath: filePath})
+          parameters = []
+          parameters.push('--syntax-check', '--no-php-ini', '--define', 'display_errors=On', '--define', 'log_errors=Off')
+          parameters.push('--file', filePath)
+          return helpers.exec(command, parameters).then (output) ->
+            messages = helpers.parse(output, @regex,
+              {filePath: filePath})
+            console.log(messages)
             resolve messages
\ No newline at end of file
```
Be advised that there's an error in your regex, You might wanna see those messages when they are logged to console.

**Comment by Arcanemagus** [comment]
It's the original regex, I didn't touch it :wink:.

I'll test those changes locally and push a commit.

**Comment by Arcanemagus** [comment]
There, that's working for me. Your diff wasn't quite correct, it had a return on the helpers.exec that was messing things up.

Do you think the @regex should just be moved down into the exec area, or left where it is?

**Comment by Arcanemagus** [comment]
Oh, and I think the "error in the regex" was that you only had -> instead of =>, so it couldn't see the @regex (which for some reason silently "works").

Can you confirm that it's working for you now?

File: lib/main.coffee [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+  provideLinter: ->
+    path = require 'path'
+    helpers = require('atom-linter')
+    child_process = require 'child_process'
+    provider =
+      grammarScopes: ['text.html.php', 'source.php']
+      scope: 'file'
+      lintOnFly: false
+      lint: (textEditor) =>
+        return new Promise (resolve, reject) =>
```
Comment (steelbrain) [inline lib/main.coffee]: How about we don't return a promise here, and return `helpers.exec` return value and add a then to it that alters it's value?

Comment (Arcanemagus) [inline lib/main.coffee]: Works for me, I'll make the change.

File: lib/main.coffee [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+      (executablePath) =>
+        @executablePath = executablePath
+
+  deactivate: ->
+    @subscriptions.dispose()
+
+  provideLinter: ->
+    path = require 'path'
+    helpers = require('atom-linter')
+    child_process = require 'child_process'
```
Comment (steelbrain) [inline lib/main.coffee]: how about you make it `require('child_process')` here and `require('path')` on it's line

Comment (Arcanemagus) [inline lib/main.coffee]: Actually, shouldn't need either of those any more.

**Comment by steelbrain** [comment]
Looks good :wink: Expect you should add EOL to files and remove the console.log

**Comment by Arcanemagus** [comment]
Fixed :+1:

**Comment by steelbrain** [comment]
Thanks :clap:

**Comment by nelson6e65** [comment]
Thanks for your work. :octocat:

**Comment by Arcanemagus** [comment]
Finally got tired of the deprecation warnings and figured I would do something about it :stuck_out_tongue:

**Pull request merged**